File: account/src/main/java/serp/project/account/ui/controller/MenuDisplayController.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.account.ui.controller;

import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import serp.project.account.core.domain.dto.request.AssignMenuDisplayToRoleDto;
import serp.project.account.core.domain.dto.request.CreateMenuDisplayDto;
import serp.project.account.core.domain.dto.request.GetMenuDisplayParams;
import serp.project.account.core.domain.dto.request.UpdateMenuDisplayDto;
import serp.project.account.core.usecase.MenuDisplayUseCase;
import serp.project.account.kernel.utils.AuthUtils;

import java.util.List;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/v1")
public class MenuDisplayController {
    private final MenuDisplayUseCase menuDisplayUseCase;

    private final AuthUtils authUtils;

    @PostMapping("/menu-displays")
    public ResponseEntity<?> createMenuDisplay(@Valid @RequestBody CreateMenuDisplayDto request) {
        var response = menuDisplayUseCase.createMenuDisplay(request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PutMapping("/menu-displays/{id}")
    public ResponseEntity<?> updateMenuDisplay(@PathVariable Long id,
            @Valid @RequestBody UpdateMenuDisplayDto request) {
        var response = menuDisplayUseCase.updateMenuDisplay(id, request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @DeleteMapping("/menu-displays/{id}")
    public ResponseEntity<?> deleteMenuDisplay(@PathVariable Long id) {
        var response = menuDisplayUseCase.deleteMenuDisplay(id);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/menu-displays/get-by-module/{moduleId}")
    public ResponseEntity<?> getMenuDisplaysByModuleId(@PathVariable Long moduleId) {
        var response = menuDisplayUseCase.getMenuDisplaysByModuleId(moduleId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/menu-displays/get-by-module-and-user")
    public ResponseEntity<?> getMenuDisplaysByModuleIdAndUserId(
            @RequestParam Long moduleId) {
        Long userId = authUtils.getCurrentUserId().orElse(null);
        var response = menuDisplayUseCase.getMenuDisplaysByModuleIdAndUserId(moduleId, userId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/menu-displays")
    public ResponseEntity<?> getAllMenuDisplays(
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer pageSize,
            @RequestParam(required = false) String sortBy,
            @RequestParam(required = false) String sortDir,
            @RequestParam(required = false) Long moduleId,
            @RequestParam(required = false) String search) {
        var params = GetMenuDisplayParams.builder()
                .page(page).pageSize(pageSize).sortBy(sortBy).sortDirection(sortDir)
                .moduleId(moduleId)
                .search(search != null ? search.trim() : null)
                .build();
        var response = menuDisplayUseCase.getAllMenuDisplays(params);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/menu-displays/assign-to-role")
    public ResponseEntity<?> assignMenuDisplaysToRole(@Valid @RequestBody AssignMenuDisplayToRoleDto request) {
        var response = menuDisplayUseCase.assignMenuDisplaysToRole(request.getRoleId(), request.getMenuDisplayIds());
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/menu-displays/unassign-from-role")
    public ResponseEntity<?> unassignMenuDisplaysFromRole(@Valid @RequestBody AssignMenuDisplayToRoleDto request) {
        var response = menuDisplayUseCase.unassignMenuDisplaysFromRole(request.getRoleId(),
                request.getMenuDisplayIds());
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/menu-displays/get-by-role-ids")
    public ResponseEntity<?> getMenuDisplaysByRoleIds(@RequestParam List<Long> roleIds) {
        var response = menuDisplayUseCase.getMenuDisplaysByRoleIds(roleIds);
        return ResponseEntity.status(response.getCode()).body(response);
    }

}

File: api_gateway/src/ui/controller/account/menu_display_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	request "github.com/serp/api-gateway/src/core/domain/dto/request/account"
	service "github.com/serp/api-gateway/src/core/service/account"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type MenuDisplayController struct {
	menuDisplayService service.IMenuDisplayService
}

func (m *MenuDisplayController) CreateMenuDisplay(c *gin.Context) {
	var req request.CreateMenuDisplayDto
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := m.menuDisplayService.CreateMenuDisplay(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *MenuDisplayController) UpdateMenuDisplay(c *gin.Context) {
	id, ok := utils.ValidateAndParseID(c, "id")
	if !ok {
		return
	}

	var req request.UpdateMenuDisplayDto
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := m.menuDisplayService.UpdateMenuDisplay(c.Request.Context(), id, &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *MenuDisplayController) DeleteMenuDisplay(c *gin.Context) {
	id, ok := utils.ValidateAndParseID(c, "id")
	if !ok {
		return
	}

	res, err := m.menuDisplayService.DeleteMenuDisplay(c.Request.Context(), id)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *MenuDisplayController) GetMenuDisplaysByModuleId(c *gin.Context) {
	moduleId, ok := utils.ValidateAndParseID(c, "moduleId")
	if !ok {
		return
	}

	res, err := m.menuDisplayService.GetMenuDisplaysByModuleId(c.Request.Context(), moduleId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *MenuDisplayController) GetAllMenuDisplays(c *gin.Context) {
	params := request.GetMenuDisplayParams{
		Page:          utils.ParseIntQuery(c, "page"),
		PageSize:      utils.ParseIntQuery(c, "pageSize"),
		SortBy:        utils.ParseStringQuery(c, "sortBy"),
		SortDirection: utils.ParseStringQuery(c, "sortDir"),
		ModuleId:      utils.ParseInt64Query(c, "moduleId"),
		Search:        utils.ParseStringQuery(c, "search"),
	}

	res, err := m.menuDisplayService.GetAllMenuDisplays(c.Request.Context(), &params)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *MenuDisplayController) AssignMenuDisplaysToRole(c *gin.Context) {
	var req request.AssignMenuDisplayToRoleDto
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := m.menuDisplayService.AssignMenuDisplaysToRole(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *MenuDisplayController) GetMenuDisplaysByModuleIdAndUserId(c *gin.Context) {
	moduleId, ok := utils.ValidateAndParseQueryID(c, "moduleId")
	if !ok {
		return
	}

	res, err := m.menuDisplayService.GetMenuDisplaysByModuleIdAndUserId(c.Request.Context(), moduleId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *MenuDisplayController) UnassignMenuDisplaysFromRole(c *gin.Context) {
	var req request.AssignMenuDisplayToRoleDto
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := m.menuDisplayService.UnassignMenuDisplaysFromRole(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (m *MenuDisplayController) GetMenuDisplaysByRoleIds(c *gin.Context) {
	roleIdsStr := c.Query("roleIds")
	if roleIdsStr == "" {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	roleIds, ok := utils.ValidateAndParseIDsQuery(c, "roleIds")
	if !ok {
		return
	}

	res, err := m.menuDisplayService.GetMenuDisplaysByRoleIds(c.Request.Context(), roleIds)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func NewMenuDisplayController(menuDisplayService service.IMenuDisplayService) *MenuDisplayController {
	return &MenuDisplayController{
		menuDisplayService: menuDisplayService,
	}
}
